core/src/main/java/io/opencensus/tags/TagKey.java

```diff
@@ -94,12 +80,112 @@
    * @throws IllegalArgumentException if the name is not valid.
    */
   // TODO(sebright): Make this public once we support types other than String.
-  static TagKey<Boolean> createBooleanKey(String name) {
+  static TagKeyBoolean createBooleanKey(String name) {
     checkArgument(StringUtil.isValid(name));
-    return new AutoValue_TagKey<Boolean>(name, TAG_BOOLEAN);
+    return TagKeyBoolean.create(name);
+  }
+
+  public abstract String getName();
+
+  /**
+   * Applies a function to the {@code TagKey} subclass. The function that is called depends on the
+   * type of the tag key. This is similar to the visitor pattern.
+   *
+   * @param stringFunction the function to call when the {@code TagKey} is a {@code TagKeyString}.
+   * @param longFunction the function to call when the {@code TagKey} is a {@code TagKeyLong}.
+   * @param booleanFunction the function to call when the {@code TagKey} is a {@code TagKeyBoolean}.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag key's type.
+   */
+  // TODO(sebright): Should we make this public in the first release?
+  public <T> T match(
+      Function<? super TagKeyString, T> stringFunction,
+      Function<? super TagKeyLong, T> longFunction,
+      Function<? super TagKeyBoolean, T> booleanFunction) {
+    return matchWithDefault(
+        stringFunction, longFunction, booleanFunction, TagUtils.<T>throwAssertionError());
+  }
+
+  /**
+   * Applies a function to the {@code TagKey} subclass. This method is like {@link #match(Function,
+   * Function, Function)}, except that it has a default case, for backwards compatibility when tag
+   * types are added.
+   *
+   * @param stringFunction the function to call when the {@code TagKey} is a {@code TagKeyString}.
+   * @param longFunction the function to call when the {@code TagKey} is a {@code TagKeyLong}.
+   * @param booleanFunction the function to call when the {@code TagKey} is a {@code TagKeyBoolean}.
+   * @param defaultFunction the function to call when the tag key has a type other than
+   *     {@code String}, {@code long}, or {@code boolean}.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag key's type.
+   */
+  // TODO(sebright): Should we make this public in the first release?
+  // TODO(sebright): How should we deal with the possibility of adding more tag types in the future?
+  //                 Will this method work? What type of parameter should we use for the default
+  //                 case? Should we remove the non-backwards compatible "match" method?
+  public abstract <T> T matchWithDefault(
```

Comment:
I added an example that shows how a user could handle different tag types differently, if they had a variable of type `TagKey`.  I'm still not sure why a user would need to call this method, since the user creates the tag keys and could assign them to variables of type `TagKeyString`, `TagKeyLong`, etc., which still have the tag type information.  I added the `TagKey.match` methods for consistency with the `Tag` class.